API: Add endpoint for fetching location data

ryelle opened the pull request:
Fixes #15080 by adding a new endpoint `/data/locations` returning all available continents, countries, and states; with options to restrict to a continent or country.

```
GET /data # returns an index describing the following endpoints
GET /data/locations # returns all continents, countries and states.
GET /data/locations/<location> # if given a continent code, return all countries and states of that continent.
GET /data/locations/<location> # if given a country code, return all states of a country.
```

This returns data that looks like the following. It uses `WC()->countries->get_*` etc, so any filtered content should be reflected in the output.

```
[
	{
		"code": "NA",
		"name": "North America",
		"countries": [
			{
				"code": "AG",
				"name": "Antigua and Barbuda",
				"states": []
			},
			…
			{
				"code": "US",
				"name": "United States (US)",
				"states": {
					"AL": "Alabama",
					"AK": "Alaska",
					…
```

To test:

- Run the unit test: `phpunit tests/unit-tests/api/data.php`
- Manually test by loading the endpoint, `/data/locations` optionally with continent & country codes, ex `/data/locations/eu`, `/data/locations/us`, or fake data to see an error `/data/locations/aa`
- `/data` lists an "index" view
- All new endpoints should be restricted to store managers

ryelle committed:
API: Add endpoints for fetching location data — continents, countries, and states — from a given site.

claudiosanches commented on includes/api/class-wc-rest-data-location-controller.php:
```diff
@@ -0,0 +1,148 @@
+ * @package WooCommerce/API
+ * @extends WC_REST_Controller
+ */
+class WC_REST_Data_Location_Controller extends WC_REST_Data_Controller {
```
Class name should be `WC_REST_Data_Locations_Controller`

claudiosanches reviewed (changes requested):
Missing schema in all classes too.

mikejolley commented on includes/api/class-wc-rest-data-controller.php:
```diff
@@ -0,0 +1,116 @@
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_items( $request ) {
+		$data    = array();
```
nitpick; alignment

mikejolley commented on includes/api/class-wc-rest-data-controller.php:
```diff
@@ -0,0 +1,116 @@
+				'description' => __( 'List of supported continents, countries, and states; restricted to the given continent', 'woocommerce' ),
+			),
+			array(
+				'slug' => 'locations/<continent>/<country>',
```
Cc'ing @justinshreve here too for feedback. How important is continent when it comes to country codes? It feels kind of weird to require it just to get a list of states for a country - all countries are unique, ISO standard codes...

justinshreve replied:
Yeah I think we could probably drop it.

I know settings will mostly be using `/data/locations` that returns everything, but in terms of other use cases where we might need specific states for a country - it does seem easier to not require the continent.

ryelle replied:
Is having just continent useful? We could just get rid of `/data/locations/<continent>` and only have `/data/locations/<country>`.

mikejolley replied:
Although existing use cases don't need it (?) I think it can be useful. It might be needed in the shipping zone implementation.

mikejolley commented on includes/api/class-wc-rest-data-location-controller.php:
```diff
@@ -0,0 +1,148 @@
+					),
+					'country' => array(
+						'description' => __( 'Country name to restrict states list', 'woocommerce' ),
+						'type'        => 'string',
```
Correct me if I'm wrong, but I believe we can define valid values for these can we not?

ryelle replied:
I'm not sure what you mean - do you mean whitelisting countries?

mikejolley commented on includes/api/class-wc-rest-data-location-controller.php:
```diff
@@ -0,0 +1,148 @@
+
+		$data = array();
+		foreach ( $continents as $continent_code => $continent_list ) {
+			if ( isset( $request['continent'] ) && ( strtolower( $continent_code ) !== strtolower( $request['continent'] ) ) ) {
```
Since codes are always upper case (?) wouldn't it be simpler to just sanitize and uppercase the user input?

ryelle replied:
Since these codes are filterable, I don't think we can rely on them always being uppercase

mikejolley commented on tests/unit-tests/api/data.php:
```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * Tests for Data API.
```
🎉

mikejolley reviewed (changes requested):
@justinshreve This could use your eyes too.

ryelle committed:
Rename files & classes to match endpoint, remove abstract class

ryelle committed:
Remove /data/locations/<continent>/<country> in favor of one endpoint which accepts either continent codes or country codes

ryelle committed:
Add schemas

justinshreve commented on includes/api/class-wc-rest-data-controller.php:
```diff
@@ -0,0 +1,126 @@
+				'slug' => 'locations',
+				'description' => __( 'List of supported continents, countries, and states', 'woocommerce' ),
+			),
+			array(
```
I'm not sure if this second should be returned here at the root.  I think the `locations` resource and in the future `currencies` would be here, but not anything one level down.

justinshreve commented on includes/api/class-wc-rest-data-locations-controller.php:
```diff
@@ -0,0 +1,146 @@
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_items( $request ) {
+		$continents = WC()->countries->get_continents();
```
Alignment needed for all 5 of these lines.

justinshreve commented on tests/unit-tests/api/data.php:
```diff
@@ -0,0 +1,114 @@
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertTrue( is_array( $locations ) );
+		$this->assertGreaterThan( 1, count( $locations ) );
+		$this->assertTrue( isset( $locations[0]['code'] ) );
```
We can use `assertNotEmpty( $locations[0]['code'])` here instead, and on other similar checks.

ryelle commented:
I think I've addressed the feedback here — renamed the classes and files, fixed my copy-paste issues 😣 , added schemas, and changed the endpoints so a continent code is not required to get a states list for one country. I'll update the main description to reflect this.

justinshreve commented on tests/unit-tests/api/data.php:
```diff
@@ -0,0 +1,114 @@
+		$response = $this->server->dispatch( new WP_REST_Request( 'GET', '/wc/v2/data' ) );
+		$index = $response->get_data();
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( 2, count( $index ) );
```
We could use `assertCount` here instead: `$this->assertCount( 2, $index );`

(I know other API tests are doing assertEquals too, but would be good to fix those in the future).

emdashcodes reviewed (changes requested):
Looking good! I left a few notes. Most minor coding standards things, but I did note a few small improvements.

justinshreve commented:
Some quick notes on responses:

When requesting a specific country (say `https://:site/wp-json/wc/v2/data/locations/CA`) I think it would make sense to drop off the continent info. So instead of 

```[
  {
    "code": "NA",
    "name": "North America",
    "countries": [
      {
        "code": "CA",
        "name": "Canada",
        "states": {
          "AB": "Alberta",
          "BC": "British Columbia",
          "MB": "Manitoba",
          "NB": "New Brunswick",
          "NL": "Newfoundland and Labrador",
          "NT": "Northwest Territories",
          "NS": "Nova Scotia",
          "NU": "Nunavut",
          "ON": "Ontario",
          "PE": "Prince Edward Island",
          "QC": "Quebec",
          "SK": "Saskatchewan",
          "YT": "Yukon Territory"
        }
      }
    ]
  }
]
```

We would return something like

```
{
        "code": "CA",
        "name": "Canada",
        "states": {
          "AB": "Alberta",
          "BC": "British Columbia",
          "MB": "Manitoba",
          "NB": "New Brunswick",
          "NL": "Newfoundland and Labrador",
          "NT": "Northwest Territories",
          "NS": "Nova Scotia",
          "NU": "Nunavut",
          "ON": "Ontario",
          "PE": "Prince Edward Island",
          "QC": "Quebec",
          "SK": "Saskatchewan",
          "YT": "Yukon Territory"
        }
}
```

When requesting something like `https://:site/wp-json/wc/v2/data/locations/NA`, the whole object should not be wrapped in an array like it is now:

```
[
  {
    "code": "NA",
    "name": "North America",
    ...
  }
]
```

(Same issue in the CA response above too).

ryelle committed:
Address code & documentation style feedback

ryelle committed:
Remove locations sub-route from data index list

ryelle committed:
Switch to more appropriate	assert* calls

ryelle commented:
How would changing the response like that affect the schema? I can see removing the array wrapper for continent & country responses, because we know we're only requesting one item, and that wouldn't affect the schema as it is now.

(Honestly I'm not sure I see the relationship of schema for this data, since we're not updating anything on this endpoint)

claudiosanches commented:
> Honestly I'm not sure I see the relationship of schema for this data, since we're not updating anything on this endpoint

Maybe better write first the schema, since it will make easy to code the data that should return.

mikejolley commented:
@justinshreve Good point from @ryelle ^ https://github.com/woocommerce/woocommerce/pull/15162#issuecomment-302824207

If the schema differs based on the input, how do you represent that in the response? To have 2 different schemas you'd need /country/ /continent/? Which I think brings this full circle 😱 

Unless the schema was abstracted somewhat?

```
code - Code of that requested
name - Name of that requested
locations (?) - List of state or country codes depending on what was requested.
```

Then it applies to both?

Let's discuss before making further changes so we have a solution.

justinshreve commented:
Yeah good point about the schema @ryelle & @mikejolley.

I like your suggestion Mike. I think returning `code`, `name`, and an array of `locations` makes sense.

ryelle commented:
So with the abstracted schema, would `/<continent>` still include the nested { code, name, list of states } in each country on that continent?

justinshreve commented:
Yeah I think we can/should still do that.

justinshreve commented:
Wouldn't it just be similar to what it is now, following the schema updates?

```
  {
    "code": "NA",
    "name": "North America",
    "locations": [
        {
           "code": "US",
           "name": "United States (US)",
           "locations": [
            {
               "code": "AL",
               "name": "Alabama",
            ...
```

justinshreve commented:
Then if you requested /US, the country part would drop off.

mikejolley commented:
Ah I get you. @ryelle can that be represented in the schema without issue?

ryelle committed:
Update response to use code/name/locations format, and only return relevant requested data

ryelle committed:
Add schema to response, updated for new response format

ryelle commented:
These latest changes updates the response object based on what filter is passed in (country, continent, or none). I've also tried to update the schema to reflect that the items at different levels may be countries, continents, states, or absent (it makes more sense if you look at the schema and some response objects).

claudiosanches commented:
I could imagine this with some other endpoints, like:

All results:

`GET /data/locations`

```json
[
  {
    "code": "AF",
    "name": "Africa",
    "countries": [...],
    "_links": {
      "self": [
        {
          "href": "https://example.com/wp-json/wc/v2/data/locations/af"
        }
      ],
      "collection": [
        {
          "href": "https://example.com/wp-json/wc/v2/data/locations"
        }
      ]
    }
  },
  {
    "code": "AN",
    "name": "Antarctica",
    "countries": [...],
    "_links": {
      "self": [
        {
          "href": "https://example.com/wp-json/wc/v2/data/locations/an"
        }
      ],
      "collection": [
        {
          "href": "https://example.com/wp-json/wc/v2/data/locations"
        }
      ]
    }
  },
  ...
]
```

Specific continent:

`GET /data/locations/af`

```
{
  "code": "AF",
  "name": "Africa",
  "countries": [...],
  "_links": {
    "self": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/af"
      }
    ],
    "collection": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations"
      }
    ]
  }
}
```

Specific country:

`GET /data/locations/af/za`

```json
{
  "code": "ZA",
  "name": "South Africa",
  "states": [
    {
      "code": "EC",
      "name": "Eastern Cape"
    },
    {
      "code": "FS",
      "name": "Free State"
    },
    {
      "code": "GP",
      "name": "Gauteng"
    },
    {
      "code": "KZN",
      "name": "KwaZulu-Natal"
    },
    {
      "code": "LP",
      "name": "Limpopo"
    },
    {
      "code": "MP",
      "name": "Mpumalanga"
    },
    {
      "code": "NC",
      "name": "Northern Cape"
    },
    {
      "code": "NW",
      "name": "North West"
    },
    {
      "code": "WC",
      "name": "Western Cape"
    }
  ],
  "_links": {
    "self": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/af/za"
      }
    ],
    "up": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/af"
      }
    ]
  }
}
```

This should require a class for `data/locations` and `data/locations/af`, and another class for `data/locations/af/za`, as well two different schema too.

But your currently schema is correct and we can go with it.
Just need to do some styling fixes, like use period in the schema and docs, like we does for the other endpoints.

ryelle committed:
Documentation style: add periods

ryelle commented:
I dropped the `/locations/<continent>/<country>` based on this comment https://github.com/woocommerce/woocommerce/pull/15162#discussion_r117454641 about real use-cases, so I'm going to leave the endpoints as-is.

I've added periods to all returned doc strings, and I _think_ I've addressed all the code & doc style feedback.

emdashcodes reviewed (approved):
Did another pass over the code. Tests pass. Responses look good.

I don't have anything else to add here. LGTM!

mikejolley commented:
First off @ryelle, really sorry for going backwards and forwards on this. I appreciate the work you're doing here and acknowledge we're lacking good guidelines for new endpoints and I'm trying to rectify that this week.

The schema of this is bothering me, in particular this part:

> List of locations in this country (absent if state).

I feel we may just be trying to do too much in that one endpoint. 

Going back to @claudiosanches comment above (https://github.com/woocommerce/woocommerce/pull/15162#issuecomment-303245008) whilst we should definitely not add the nesting back for continents (because it would be hard to consume) we could make a minor change here and have 2 separate endpoints.

So this structure:

- `data/locations/countries/`
   - `data/locations/countries/gb`
   - `data/locations/countries/us`
   - ...
- `data/locations/continents/`
   - `data/locations/continents/eu`
   - `data/locations/continents/na`
   - ...

`data/locations` will just list resources like your `data` endpoint.

So a specific continent looks like:

```
{
  "code": "EU",
  "name": "Europe",
  "countries": [...],
  "_links": {
    "self": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/continents/af"
      }
    ],
    "collection": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/continents"
      }
    ]
  }
}
```

And a specific country looks like:

```
{
  "code": "ZA",
  "name": "South Africa",
  "states": [
    {
      "code": "EC",
      "name": "Eastern Cape"
    },
    {
      "code": "FS",
      "name": "Free State"
    },
    {
      "code": "GP",
      "name": "Gauteng"
    },
    {
      "code": "KZN",
      "name": "KwaZulu-Natal"
    },
    {
      "code": "LP",
      "name": "Limpopo"
    },
    {
      "code": "MP",
      "name": "Mpumalanga"
    },
    {
      "code": "NC",
      "name": "Northern Cape"
    },
    {
      "code": "NW",
      "name": "North West"
    },
    {
      "code": "WC",
      "name": "Western Cape"
    }
  ],
  "_links": {
    "self": [
      {
        "href": "https://example.com/wp-json/wc/v2/data/locations/countries/za"
      }
    ]
  }
}
```

Does that sound acceptable? @claudiosanches 

I'll be talking to Todd and @mattyza later about what version these endpoints can realistically go in, since we're going to need to version bump the API for discovery. With that in mind, if you could add a `feature/rest-api-v3` branch and target that instead of `master` we'll be able to approve and merge this until a decision is made. 

cc @justinshreve

claudiosanches commented:
@mikejolley sounds good.

Countries inside the `data/locations/continents` could even not return any extra information, maybe only codes and paths to help query in the data/locations/continents/ISO-CODE` endpoint.

ryelle commented:
I don't really know if we need another index route, what do you think of switching down to just `/data/continents` and `/data/countries`?

ryelle committed:
Switch endpoints from `/data/locations` into two, continents and countries.

ryelle committed:
Add relation links to responses

ryelle committed:
Update location -> continents/countries in keys, filter names, documentation

ryelle commented:
Added a new branch `feature/rest-api-v3` and retargeted this one for that branch. I also switched the endpoints to `/data/countries` & `/data/continents`, updated the schema, tests, & documentation to reflect that, and added relation links to all responses.

claudiosanches commented:
@ryelle still missing handle each single item as an individual result, so you'll be able to set `_links` as @mikejolley described.
Check our other REST API classes, there we have `get_item()` and `get_items()` to solve it.

ryelle committed:
Refactor to use `get_item` for single continents/countries, with a get_* helper function. This also adds the `_links` data to each item in the collection.

mikejolley commented:
@claudiosanches Can you be a little more detailed? What is missing? Is it there now?

claudiosanches commented:
@mikejolley Kelly already fixed it, but I was talking about treating each element as an element in individual, so including `_links` with `self` and `collection`.

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,285 @@
+		 * @param array            $data     The original list of continent(s), countries, and states.
+		 * @param WP_REST_Request  $request  Request used to generate the response.
+		 */
+		return apply_filters( 'woocommerce_rest_prepare_data_continent', $response, $continent, $request );
```
This filter should be included into `prepare_item_for_response`.

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,285 @@
+		}
+
+		$continent['countries'] = $local_countries;
+		$response = $this->prepare_item_for_response( $continent, $request );
```
```php
$response = $this->prepare_item_for_response( $continent, $request );
$response = $this->prepare_response_for_collection( $response );
```

Both lines should be part of `get_item` and `get_items` and not while getting the "row data".

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,285 @@
+		if ( ! is_wp_error( $item ) ) {
+			$response->add_links( $this->prepare_links( $item ) );
+		}
+		return $response;
```
Missing `woocommerce_rest_prepare_*` filter.

claudiosanches commented on includes/api/class-wc-rest-data-countries-controller.php:
```diff
@@ -0,0 +1,237 @@
+			}
+		}
+		$country['states'] = $local_states;
+		$response = $this->prepare_item_for_response( $country, $request );
```
Need the same change here, putting both into `get_item` and `get_items`.

claudiosanches commented on includes/api/class-wc-rest-data-countries-controller.php:
```diff
@@ -0,0 +1,237 @@
+	public function prepare_item_for_response( $item, $request ) {
+		$response = rest_ensure_response( $item );
+		$response->add_links( $this->prepare_links( $item ) );
+		return $response;
```
Missing `woocommerce_rest_prepare_*` filter.

claudiosanches commented on includes/api/class-wc-rest-data-controller.php:
```diff
@@ -0,0 +1,156 @@
+
+		// Wrap the data in a response object.
+		$response = rest_ensure_response( $data );
+		$response->add_links( array(
```
This class should have a `prepare_links` method like all other classes in our REST API.

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,285 @@
+	 */
+	protected function prepare_links( $item ) {
+		$links = array();
+		if ( array_key_exists( 'code', $item ) ) {
```
Not sure if need to check for `array_key_exists( 'code', $item )` since now all items should have `self` and `collection`.

ryelle committed:
Move item-prepare code into the prepare method

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,279 @@
+	public function get_item( $request ) {
+		$data = $this->get_continent( strtoupper( $request['location'] ), $request );
+		if ( empty( $data ) ) {
+			$data = new WP_Error( 'woocommerce_rest_data_invalid_location', __( 'There are no locations matching these parameters.', 'woocommerce' ), array( 'status' => 404 ) );
```
No need to pass `WP_Error` to variables, you can just use `return` and stop the code here.
This will avoid the need to check for `is_wp_error` in `prepare_item_for_response` too.

claudiosanches commented on includes/api/class-wc-rest-data-continents-controller.php:
```diff
@@ -0,0 +1,279 @@
+	 * @return WP_REST_Response $response Response data.
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+		$response = rest_ensure_response( $item );
```
Can change it to:

```php
		$data     = $this->add_additional_fields_to_object( $item, $request );
		$response = rest_ensure_response( $data );
		$response->add_links( $this->prepare_links( $item ) );
```
 
`add_additional_fields_to_object` will allow extend results using `register_rest_field`.

mikejolley commented:
@justinshreve When those last items are done, are you going to cherry-pick this into the wc-api-dev repo?

ryelle committed:
Return wp_error directly, and allow for extending endpoints via `register_rest_field`

justinshreve commented:
@mikejolley Yeah, I can get these into the plugin.

Are we good to go?

claudiosanches reviewed (approved):
Looks good, great work :)

justinshreve commented:
Merged into the feature plugin over at https://github.com/woocommerce/wc-api-dev/pull/3. Thanks @ryelle!

I think we can close this, since we are not going to be using the v3 branch.
